sdk/src/prices/calculatePoolPrices.ts

```diff
@@ -0,0 +1,293 @@
+import { AddressUtil, DecimalUtil, Percentage } from "@orca-so/common-sdk";
+import { Address, BN, translateAddress } from "@project-serum/anchor";
+import { u64 } from "@solana/spl-token";
+import { PublicKey } from "@solana/web3.js";
+import Decimal from "decimal.js";
+import {
+  DecimalsMap,
+  defaultConfig,
+  defaultThresholdConfig,
+  GetPricesConfig,
+  PoolMap,
+  PriceMap,
+  ThresholdConfig,
+  TickArrayMap,
+  PoolObject,
+} from ".";
+import { swapQuoteWithParams } from "../quotes/public/swap-quote";
+import { TickArray, WhirlpoolData } from "../types/public";
+import { PoolUtil, PriceMath, SwapUtils } from "../utils/public";
+import { PDAUtil } from "../utils/public/pda-utils";
+
+/**
+ * calculatePoolPrices will calculate the price of each token in the mints array.
+ * Each token will be priced against the first quote token in the config.quoteTokens array
+ * with sufficient liquidity. If a token does not have sufficient liquidity against the
+ * first quote token, then it will be priced against the next quote token in the array.
+ * If a token does not have sufficient liquidity against any quote token,
+ * then the price will be set to null.
+ * The threshold for "sufficient liquidity" is defined by the thresholdConfig
+ *
+ * The caller of the function must provide the accounts through the following parameters:
+ *  - poolMap: A map of pool addresses to pool data
+ *  - tickArrayMap: A map of pool addresses to tick array data
+ *  - decimalsMap: A map of token mint addresses to token decimals
+ *
+ * fetchPoolPrices.ts provides functions to fetch these accounts
+ */
+export function calculatePoolPrices(
+  mints: PublicKey[],
+  poolMap: PoolMap,
+  tickArrayMap: TickArrayMap,
+  decimalsMap: DecimalsMap,
```

Comment:
note that this namespace will can also host functions related to getting prices for a specific pool.

Reply:
Another thing that's popping out to me is that we should be re-using pool-graph that's being used on smart-router to do the walk.

The UI already needs to generate this graph once to support routing. We can save some computation if we want to support prices on top of that.

Reply:
> [side note for future] Instead of storing decimals only, can we use the types defined token package that @terranmoccasin is building? Having a mapping from token -> price can also be useful since users may want to present the info as "SYMBOL - PRICE" instead of "MINT - PRICE"

potentially! although IIRC, the token package will not just fetch decimal data but metadata from metaplex accounts and other sources. we may not want to require callers of `PriceModule` to also fetch all of this additional data just to calculate decimals

on the other hand, reusable code to fetch decimals that's shared by both the token package and the price module sounds like a reasonable idea

Reply:
> the token package will not just fetch decimal data but metadata from metaplex accounts and other sources.

Using the current `TokenFetcher` implementation, it's actually possible to fetch only the on-chain mint account data. You could create a `TokenFetcher` that doesn't have any `MetadataProvider`s added.

https://github.com/orca-so/orca-sdks/blob/main/packages/token-sdk/src/token-fetcher.ts#L40

Fetcher always grabs mint account data first before supplementing the returned `Token` object with optional metadata.